File: codes/typescript/chapter_tree/binary_tree_bfs.ts
/**
 * File: binary_tree_bfs.ts
 * Created Time: 2022-12-14
 * Author: Justin (xiefahit@gmail.com)
 */

import { type TreeNode } from '../module/TreeNode';
import { arrToTree } from '../module/TreeNode';
import { printTree } from '../module/PrintUtil';

/* 层序遍历 */
function hierOrder(root: TreeNode | null): number[] {
    // 初始化队列，加入根结点
    const queue = [root];
    // 初始化一个列表，用于保存遍历序列
    const list: number[] = [];
    while (queue.length) {
        let node = queue.shift() as TreeNode; // 队列出队
        list.push(node.val); // 保存结点
        if (node.left) {
            queue.push(node.left); // 左子结点入队
        }
        if (node.right) {
            queue.push(node.right); // 右子结点入队
        }
    }
    return list;
}

/* Driver Code */
/* 初始化二叉树 */
// 这里借助了一个从数组直接生成二叉树的函数
var root = arrToTree([1, 2, 3, 4, 5, 6, 7]);
console.log('\n初始化二叉树\n');
printTree(root);

/* 层序遍历 */
const list = hierOrder(root);
console.log('\n层序遍历的结点打印序列 = ' + list);

File: codes/ruby/chapter_tree/binary_tree_bfs.rb
=begin
File: binary_tree_bfs.rb
Created Time: 2024-04-18
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

require_relative '../utils/tree_node'
require_relative '../utils/print_util'

### 层序遍历 ###
def level_order(root)
  # 初始化队列，加入根节点
  queue = [root]
  # 初始化一个列表，用于保存遍历序列
  res = []
  while !queue.empty?
    node = queue.shift # 队列出队
    res << node.val # 保存节点值
    queue << node.left unless node.left.nil? # 左子节点入队
    queue << node.right unless node.right.nil? # 右子节点入队
  end
  res
end

### Driver Code ###
if __FILE__ == $0
  # 初始化二叉树
  # 这里借助了一个从数组直接生成二叉树的函数
  root = arr_to_tree([1, 2, 3, 4, 5, 6, 7])
  puts "\n初始化二叉树\n\n"
  print_tree(root)

  # 层序遍历
  res = level_order(root)
  puts "\n层序遍历的节点打印序列 = #{res}"
end
